src/snowflake/connector/nanoarrow_cpp/ArrowIterator/CArrowTableIterator.cpp

```diff
@@ -305,10 +304,10 @@ void CArrowTableIterator::convertScaledFixedNumberColumnToDecimalColumn_nanoarro
         SF_CHECK_ARROW_RC(returnCode, "[Snowflake Exception] error appending null to arrow array, error code: %d", returnCode);
     } else {
         auto originalVal = ArrowArrayViewGetIntUnsafe(columnArray, rowIdx);
-        std::shared_ptr<ArrowDecimal> arrowDecimal = std::make_shared<ArrowDecimal>();
-        ArrowDecimalInit(arrowDecimal.get(), 128, 38, scale);
-        ArrowDecimalSetInt(arrowDecimal.get(), originalVal);
-        returnCode = ArrowArrayAppendDecimal(newArray, arrowDecimal.get());
+        ArrowDecimal arrowDecimal;
+        ArrowDecimalInit(&arrowDecimal, 128, 38, scale);
+        ArrowDecimalSetInt(&arrowDecimal, originalVal);
+        returnCode = ArrowArrayAppendDecimal(newArray, &arrowDecimal);
```

Comment:
in this case won't `arrowDecimal` be GCed as it's a local variable?

Reply:
There is no GC in C++. Instead, the memory pattern is well specified:

* Previously, when `std::shared_ptr<ArrowDecimal> arrowDecimal` was created, we would allocate some memory in the heap to back the `arrowDecimal`. When it went out of scope, `shared_ptr` would decrement the reference count, it would hit 0, and we would free the memory that backed the `arrowDecimal`.
* Now, when `ArrowDecimal arrowDecimal` is created, we allocate memory on the stack. When it goes out of scope, the relevant stack memory is freed.

Allocating memory on the stack is more-or-less "free". This variable probably has a fixed location in the stack, so the compiler can emit assembly that just uses `ebp - fixedNumber` for `arrowDecimal`.

Allocating memory on the heap is a lot more complex.

Also, if `ArrowArrayAppendDecimal` _does_ require the variable to stick around, then there's already a bug in the code - since the `shared_ptr` is freed, there is no guarantee that the memory would stay valid.

Reply:
thanks for refreshing my memory on c++, yeah you are right.
the change looks good to me